### DISCUSSION TITLE
Setting 'main' Element event handlers after creation

### DISCUSSION THREAD

**Discussion by petergaultney** [zauberzeug/nicegui#2687] [Ideas / Feature Requests]
For context, this is asking for something somewhat like the opposite of https://github.com/zauberzeug/nicegui/discussions/2538

Many elements have a built-in event handler, like `on_click` for Button, or `on_change` for ValueElement. These cannot be registered after the fact using `.on(...)` because they operate on the level of the NiceGUI framework (see side note below) rather than the underlying Quasar/Vue framework - in other words, they expose some state that's internal to the Python object itself, like the `BindableProperty` `.value` on ValueElement, which is itself derived from the Quasar state internally using `.on(...)`.

There are several reasons someone might want this - as a general concept, you might think of it as "separating UI from behavior". I'd like to be able to construct my UI in one place and then 'bind' a bunch of behavior in a separate place.

A specific case is the case where I want to have a button that 'saves' data from various inputs, and I want to use the following functional-ish style to do so. 

```py
def render_form(..., db_id: int):
    # UI comes first
    # and the button is first of all because I want it to sit at the top/left of the screen
    save_button = ui.button('save')
    is_cold = ui.checkbox(...)
    color = ui.input(...)
    name = ui.input(...)
    
    def save():
        database.save(db_id, is_cold=is_cold.value, name=name.value, color=color.value)

    save_button.on_click = save
```

if all the inputs showed up in the UI 'before' the button, i could easily do `ui.button('save', on_click=save)` - but they don't. The fact that the layout of my UI affects how I can arrange my non-layout code is surprising!

In Python, of course, `save` would _work_ if defined prior to the Elements - but this is both somewhat non-intuitive for the average programmer (why am I allowed to reference things that don't 'exist' yet?) and also unpleasant from the perspective of something like `mypy`, which refuses to infer the types of arguments 'closed over' by the `save` function if they have not yet been defined. Moving `save` below the element definitions as in the above code solves this for mypy nicely.

I've run into variations of this problem with `on_change` as well, where I really want to keep my UI definition separate from the code that hooks up to the inputs, listens for changes, potentially enables/disables the Save button based on validity, etc.

You _can_ use `ValueElement.bind_value_to` "after the fact", but this only works for ValueElement (not for Button), and if all you want is a callback when the value changes, then it's quite inconvenient to have to create a proxy object with a settable `value` property just so you can receive that callback. A much more pleasant callback definition pattern would look something like creating a simple closure around the element, however this is not possible if `on_change` cannot be defined after the creation of the ValueElement like so:

```py
def render_ui(...):
    input_a = ui.input(...)
    input_b = ui.input(...)

    # print value when changed:
    input_a._on_change = lambda: print(f"input a changed to: {input_a.value}")
    input_b._on_change = lambda: print(f"input b changed to: {input_b.value}")
```

(The above actually works - it's just ugly because it's using part of ValueElement that is not public and therefore not supported/guaranteed to work later on.)

It's quite possible there are other ways of accomplishing this in NiceGUI that I'm missing. Overall, I _suspect_ that other users will probably try (and fail) to do the same thing I'm doing - it's not intuitive to me why the 'blessed' event handler for Elements would not be something you could create after the fact just like you can do with the lower-level handlers (e.g. `.on('update:value')`) or things like `.classes` or `.props`. In fact, these high-level handlers are about the _only_ thing that doesn't support the "modify the element after creation" pattern that most of the rest of the Element API seems to follow.

---- 

Side note: I had to do a fair amount of digging through source code to understand the fundamental 'difference' between, e.g., `ValueElement(on_change=x)` and the `ValueElement.on` method.  In fact I was initially foolish enough to think that `.on('change')` would somehow be the same thing as `on_change` - and the event _does_ exist, but it fires under quite different circumstances from what NiceGUI uses for it's own `ValueElement.on_change`. 

The documentation did not _seem_ to explain the fact that `on_change` (or `on_click`) relies on the NiceGUI python object's state, including pre-running certain things like `_value_to_model_value` on the value as received from the specific _Quasar_ event `update:value`, whereas `.on` is looking at the 'raw' HTML/JS events.  I may very well have missed the relevant documentation, but I kind of think it might be worth adding some additional language to the section [here](https://nicegui.io/documentation/section_action_events#generic_events) to call out specifically that the "predefined events" may contain additional NiceGUI-specific logic, such as value-parsing, and that therefore the underlying JS events are lower-level and not interchangeable with the NiceGUI ones.

I know it's very hard to satisfy everybody with docs, and I'm okay with the hour or two that I spent digging into the code to figure this out (your code is pretty dang readable, which is a huge plus for `nicegui` as a framework!). So this is just a side note. :)

**Comment by falkoschindler** [comment]
Thanks for the suggestion, @petergaultney!

You're right. We should add the possibility to assign event handlers after instantiation, like adding style or binding. For events like a button click this should be straightforward, because the initializer simply calls `.on()`.

But there are cases where the current implementation is a bit different. The `ValueElement` calls `.on()` to get value updates, but holds the change handler in a member variable. In order for `.on_value_change()` to work when called multiple times, we probably need to hold a _list_ of change handlers.

I just started the branch https://github.com/zauberzeug/nicegui/tree/event-registration where I sketched a solution for `ui.button` and value elements. The following elements are still missing: color_picker, echart, interactive_image, joystick, json_editor, keyboard, list, menu_item, scene, scroll_area, table, tree, and upload. Would you like to contribute? 🙂

**Reply by petergaultney** [reply]
i would _love_ to have the ability to have multiple of these, yes! Thank you for thinking of that. It occurred to me that I _wanted_ this while I was writing up this issue, but it didn't occur to me that it would be a better idea to use the builder pattern, and that that meant we'd _have_ to support multiple.

Yes, I will try to carve out some time today to at least look at your changes and see what I can do to help. Hopefully the rest aren't way more complicated than the ones you've done ;)

**Reply by petergaultney** [reply]
see PR #2708 !

**Comment by petergaultney** [comment]
this is closed by #2704 and I believe will be part of version 1.4.20.
